game-core/src/main/java/games/strategy/triplea/ui/MacOsIntegration.java

```diff
@@ -0,0 +1,129 @@
+import games.strategy.util.function.ThrowingConsumer;
+import lombok.extern.java.Log;
+
+/**
+ * Utility class to hide macOS integration.
+ *
+ * <p>
+ * This class uses reflection to avoid Mac-specific classes being required for compiling on other platforms.
+ * </p>
+ */
+@Log
+public final class MacOsIntegration {
+  private MacOsIntegration() {}
+
+  /**
+   * Adds the specified about handler to the application.
+   */
+  public static void addAboutHandler(final Runnable handler) {
+    checkNotNull(handler);
+
+    addHandler(
+        getHandlerClassName("java.awt.desktop.AboutHandler", "com.apple.eawt.AboutHandler"),
+        "handleAbout",
+        "setAboutHandler",
+        args -> handler.run());
+  }
+
+  private static String getHandlerClassName(final String java9OrLaterClassName, final String java8ClassName) {
+    return isJavaVersionAtLeast9() ? java9OrLaterClassName : java8ClassName;
+  }
+
+  private static boolean isJavaVersionAtLeast9() {
+    return isJavaVersionAtLeast9(SystemProperties.getJavaSpecificationVersion());
+  }
+
+  @VisibleForTesting
+  static boolean isJavaVersionAtLeast9(final String encodedJavaSpecificationVersion) {
+    try {
+      return Float.parseFloat(encodedJavaSpecificationVersion) >= 9.0F;
+    } catch (final NumberFormatException e) {
+      log.log(Level.SEVERE, "Malformed Java specification version: '" + encodedJavaSpecificationVersion + "'", e);
+      return false;
+    }
+  }
+
+  private static void addHandler(
+      final String handlerClassName,
+      final String handlerMethodName,
+      final String addHandlerMethodName,
+      final ThrowingConsumer<Object[], Exception> handler) {
+    try {
+      addHandler(newApplication(), handlerClassName, handlerMethodName, addHandlerMethodName, handler);
+    } catch (final ReflectiveOperationException e) {
+      log.log(Level.SEVERE, "Failed to add macOS application handler: " + handlerClassName, e);
+    }
+  }
+
+  @VisibleForTesting
+  static void addHandler(
+      final Object application,
+      final String handlerClassName,
+      final String handlerMethodName,
+      final String addHandlerMethodName,
+      final ThrowingConsumer<Object[], Exception> handler) throws ReflectiveOperationException {
+    final Class<?> handlerClass = Class.forName(handlerClassName);
+    final Object handlerProxy = Proxy.newProxyInstance(
+        MacOsIntegration.class.getClassLoader(),
+        new Class<?>[] {handlerClass},
+        (proxy, method, args) -> {
+          if (handlerMethodName.equals(method.getName())) {
+            handler.accept(args);
+          }
+          return null;
+        });
+    application.getClass()
+        .getDeclaredMethod(addHandlerMethodName, handlerClass)
+        .invoke(application, handlerProxy);
+  }
+
+  private static Object newApplication() throws ReflectiveOperationException {
+    return Class.forName("com.apple.eawt.Application")
+        .getConstructor((Class[]) null)
```

Comment:
Also I recall having used `#getDeclaredConstructor` in all other cases, not sure what's the difference.

Reply:
The non-`Declared` variants only look for members with public visibility.  While the `Declared` variants can find non-public members, those members tend to only be usable if you make them accessible via `setAccessible(true)`, which probably isn't going to work in Java 9+.

For consistency, I'll try to convert the other `getDeclaredMethod()` call to just `getMethod()`.

Reply:
d4fb91857